### ISSUE TITLE
postprocessor of agent

### ISSUE THREAD

**Issue by fakesecret** [RasaHQ/rasa_core#1217]
**Rasa Core version**: 0.10.1

**Python version**:  3.6

**Operating system** (windows, osx, ...): Linux

**Issue**:

As I first mentioned in https://github.com/RasaHQ/rasa_core/issues/258, while we can have preprocessor, more advanced users/programmers may need to postprocess NLU-parsed data for Database-aided checking, context-based modification, etc. An simple example is to check NLU results with a drug database, and then correct NLU results (for missing ones and misinterpreted entities) before they are sent to tracker and a decision of action is made.

To implement this functionality, it is relatively simple. We need to add an argument `message_postprocessor=None` in `Agent.handle_message`, `handle_channel` and `_create_processor`; and `MessageProcessor.__init__`. Most important, 2 lines need to be added in `MessageProcessor._handle_message_with_tracker`:
```python3
        parse_data = self._parse_message(message)
        if self.message_postprocessor is not None:  # added
            parse_data = self.message_postprocessor(parse_data, tracker) # added
```
The postprocessor is like:
```python3
def postprocess(parse_data, tracker):
    # type: (Dict, DialogueStateTracker) -> Dict
```
I already had a python module for `MessageProcessorWithPost(MessageProcessor)` and `AgentWithPost(Agent)` in order to postprocess NLU-parsed data, without modifying Rasa-core source code. It works well. If devs are interested, I can send it to you or make a PR by myself.

**Comment by akelad** [comment]
Thanks for the suggestion, just wondering why you wouldn't handle this kind of stuff in a custom action?

**Comment by akelad** [comment]
Yes that's how we would also handle this. I don't like the idea of hard coding things like this

**Comment by fakesecret** [comment]
For me, the problem is in `processor.py` that it preprocess msg, do NLU and directly predict and exucute the next action. I don't see a straightforward and clean way to work around this problem by implementing custom actions.
```python3
    def handle_message(self, message):
        # type: (UserMessage) -> Optional[List[Text]]
        # ......
        if self.message_preprocessor is not None:
            message.text = self.message_preprocessor(message.text)
        tracker = self._get_tracker(message.sender_id)
        self._handle_message_with_tracker(message, tracker)
        self._predict_and_execute_next_action(message, tracker)
        
```
@souvikg10 Thanks for your advice. In this case, you will use "postprocessing action" systematically for all intents received, and then inject the modified parsed results into core so that it make a decision for the right action, am I right? Can you elaborate a little more how you can develop a pipeline like "rasa-NLU --> postprocessing action --> rasa-core
@akelad, Thank you for your comments. What i suggested, it's to open a postprocessor door to developpers, just like preprocessor. But I also kind of agree that postprocessor is doing work more like actions.

**Comment by souvikg10** [comment]
Well for us, 
Let's take the example i gave above

our story looks like
```
* transfer_money{amount_of_money:"20"}
- set_currency
- slot{"currency":"euros"}
- utter_confirm_amount
* affirm
- make_transfer
```
Now as you know with ner_duckling, it sends back an entity - "amount-of-money" with value - 20 and then some additional info like currency 'euros' for example. 

With rasa core automatic slot filling from entity, it will set the money to the value 20 but that is not enough to make a transfer so we add a custom action set_currency that runs separately which basically parses the tracker object and looks for the entity amount-of-money and tries to parse the additional info to look for currency.
we then return SlotSet("currency":"euros"}, where currency is text slot which enables the check if this particular slot is filled, it will call the next action - utter_confirm_action which has a template
you want to send {amount-of-money} {currency} ? 
and the story goes on. So really we did some post processing outside the rasa stack because it is really linked to particular use-case, you should handle everything specific outside of rasa stack.

**Comment by souvikg10** [comment]
What you are saying is really linked to intent classification so i think you can add it as part of the config pipeline in Rasa NLU, that takes intent prediction results as input, do post processing and returns the value to Rasa core. A typical post-processor for an agent (a rasa core agent) would be after action prediction.

**Comment by fakesecret** [comment]
@souvikg10 thx for your patience. I realized that "post-processor for an agent" is ambiguous to you. What I want, is a NLU-postprocessor, but cannot be implemented in rasa-NLU alone or any other NLU software. Because I want to use Rasa core tracker to improve intent/entities prediction. So it is not inside NLU, but really one intermediate step between NLU and rasa-core action decision making.

**Comment by souvikg10** [comment]
Then leads back to the same point, that you can add an action upfront that does postprocessing after every intent detected in your stories.
* intent
- post_processing_step - return a ``` dispatcher.utter_message("/your_improved_intent") ```
* improved_intent
- your_action

That's what we do. You can make sure this step will occur with Memoization policy

Alternatively,
You can also add code in processor.py to include another step in the series of actions before saving the object in the tracker, however bear in mind this can change your solution architecturally and in my mind, it is not a good design pattern. 

Imo, a chatbot here is a state machine, your states are decided based on what the user is saying. So really a postprocessing as you call it,  is a consequence(an action) of what the machine understood from the user and what it should do to reply back to the user to go to a different state. In order to do that it needs to do some processing(maybe call a Database, maybe look back at the tracker to determine an appropriate understanding of the conversation, maybe add a slot). Even your NLU post processing is a consequence of what the machine understood from the user. every consequence is an action. 

You have both ways to improve your action prediction, by writing code in processor.py or by adding the step as part of the story. I prefer the second way as in the pattern of how rasa core is built , this makes more sense to me

**Comment by fakesecret** [comment]
@souvikg10 I see more clearly your thinking now. Because I have a requirement of postprocessing for all my intents, so I wondered if it is better to implement this step directly into processor.py, which made sense to me but much less to you. But your aguments are quite convincing, thank you. 

In your design, the story gets quite complicated, no? All 1st NLU processed intents are followed by postprocessing action, which will give verified intents. Verified intents can then be predicted by rasa-core for real actions. Am I right? postprocess action will also eliminate Intent-get-info from tracker. The order is 
```
* Intent-get-info
- postprocess 
* verified-intent-get-info
- action-get-info
```
But real stories should be all "raw intent + postprocess" pairs, plus stories like below:
```
# story 1
* verified-intent-hello
- action-hello
* verified-intent-get-info
- action-get-info
* verified-intent-bye
- action-bye
```

**Comment by souvikg10** [comment]
It depends on your problem indeed, but imagine a multi domain chatbot where postprocessing makes sense in one while it doesn’t in another. A generalistic postprocessing step wouldn’t help all cases. Even preprocessers can help to a general extent like spell checker or language detector but it wouldn’t really help if you add a pre processing step where you initialise the tracker with a particular slot because it helps in predicting a next action. 

It really depends on your use case and scalability of your use case. Would you always need this post processing to improve your NLU prediction? 

There isn’t a question of real or fake stories. It really is what a bot needs to do given it understands what the user is saying. Even if that means improving your understanding better it is still something the bot is doing.

**Comment by fakesecret** [comment]
I totally agree with your last comment. My intention is to provide a solution for this "postprocessing" problem. It seems your solution is better for more general uses, and with no requirement to changing rasa-core code.
My question now is "if I use custom actions for postprocessing as you propose, how can I design stories". In my solution, i don't need add any postprocessing action in my story; so the story file remains the same as what i had before implementing postprocessing. In your solution, you need to modify sotries. Can you show me any Rasa doc/exampe concerning this or a toy example you created? Thank you.

**Comment by souvikg10** [comment]
you just need to add a custom action in the stories. I am not sure if it helps, i am still working on this project 
https://github.com/souvikg10/rasa-core-experiment/blob/master/data/Demo-Bot/story/story.md

here i have a story when a user orders something, each time i check if the item is available or not.
then in the custom action, i check if the item is there and then i return a slot 
you can even force a new action based on the corrected intent, this don't require any stories

https://rasa.com/docs/core/api/events/#force-a-followup-action

**Comment by fakesecret** [comment]
Thx, what I don't quite understand is how to implement the following situation you previously mentioned if an intent could be changed:
```
* intent
- post_processing_step - return a ``` dispatcher.utter_message("/your_improved_intent") 
* improved_intent
- your_action
```
In the 2nd line, the chatbot utters a magic command to user, right? But this kind of magic command ("/intent") only works when it's a user message, right?

**Comment by souvikg10** [comment]
Indeed, i just noticed the issue, also remembered that before we did this outside of rasa core using botkit by adding new intents to rasa core. 

I guess you can send a UserUttered event from your custom action

https://rasa.com/docs/core/api/events/#user-sent-message

**Comment by fakesecret** [comment]
Em ... then we come back to the very first problem. UserUttered will create a side effect, as this turn will be saved into the tracker, which will pollute my deep learning (LSTM, etc.) decision making for actions. I know I also have a choice to delete/mutate the tracker... In this case, I can set story like below. For all intents I want to postprocess with (1 to N), I create all pairs. And "real story" do not contain any " raw intent" from 1 to N.
```
# story 1
* intent_1
- post_processing_step ( UserUttered: "/verified_intent1" if not corrected) 
# story N
* intent_N
- post_processing_step ( UserUttered: "/verified_intentK" if corrected) 

# real story 1
* verified_intent1
- action1
* verified_intent2
- action2
* verified_byebye
- utter_byebye
```

**Comment by fakesecret** [comment]
@souvikg10 if you want, we can chat in gitter (where I left you a msg) or somewhere else in order not to spam here ;) After our discussion, I will edit this post and close it.

**Comment by souvikg10** [comment]
Yes sure

**Comment by ctrado18** [comment]
@fakesecret What do you mean by intent correction? Can you make an example? Do you trigger an intent based on some found words (not entitities)?

**Comment by fakesecret** [comment]
@ctrado18 Hello, I want a post-processing step to correct NLU results, which include entities and intent. While entity correction is more straightforword, such as slot checking and database mapping, etc., intent correction is a little more tricky. However, you can imagine some methods using manually-crafted rules, machine learning or some sophisticated ways (re-run, modified bag-of-words, re-ranking).
An example I showed above is 
```
User: Give me info of You&Me restaurant. (NLU: entity: restaurant: You&Me; intent: get_info.restau)
Chatbot: Here is the info for this restaurant.
User: Compare You&Me and EatMe. (NLU: no entity; intent: compare.films)
 -- [postprocessor]: You&Me mapped as restaurant by slot; EatMe mapped as restaurant by database; intent corrected as compare.restau
Chatbot: here is the restaurant comparison results: blabla
```
For the 2nd request, the intent was first interpreted as compare.films, which was incorrect, without any entity extracted. My postprocessor found "You&Me" was a restaurant because of the previous conservation, and "EatMe" was also one based on database mapping; Re-evaulation of compare.films found that it was less probable given 2 entities of "restaurant", and the postprocessor re-ranked intents and suggested compare.restau, which turns to be correct. Following these corrected entities and intent, chatbot at last gave a comparison between these 2 restaurants.
